# 240312

러스트 정리 (3)

## 추리게임 만들기

https://doc.rust-kr.org/ch02-00-guessing-game-tutorial.html


### 1. Rust 프로젝트 생성


```rust
// Cargo.toml
[dependencies]
rand = "0.8.5"


// main.rs
use std::{cmp::Ordering, io};
use rand::Rng;

fn main() {
    println!("Guess the number!");

    let secret_number = rand::thread_rng().gen_range(1..=100);

    loop {
        println!("Please input your guess.");

        let mut guess = String::new();

        io::stdin()
            .read_line(&mut guess)
            .expect("Failed to read line");

        let guess: u32 = match guess.trim().parse() {
            Ok(num) => num,
            Err(_) => continue,
        };

        println!("You guessed: {guess}");

        match guess.cmp(&secret_number) {
            Ordering::Less => println!("Too small!"),
            Ordering::Greater => println!("Too big!"),
            Ordering::Equal => {
                println!("You win!");
                break;
            }
        }
    }
}
```
- 함수는 fn 키워드 사용
- 내부, 외부 라이브러리(크레이트)는 use 키워드 사용
- 반복문 키워드 loop 있음(while,for도 있음)
- match를 사용해서 조건에 맞는 기능 수행
- expect로 에러 처리

### 2. Python 프로젝트 생성

```python
from random import randint

def main():
    print("Guess the number!")

    secret_number = randint(1,100)
    print(secret_number)
    while True:
        print("Please input your guess.")

        guess = input()
        if guess.isdigit():
            guess = int(guess) 
        else:
            continue

        print("You guessed: {guess}")

        if guess > secret_number:
            print("Too big!")
        elif guess < secret_number:
            print("Too small!"),
        else: # guess == secret_number:
            print("You win!")
            break


if __name__ == "__main__":
    main()
```
- 최대한 비슷하게 하려곤 했지만 귀찮음이 포함되어버렸다.
- 외부 라이브러리 혹은 내장함수 등을 사용할때 import 사용. 
- 에러 처리 없음

### 3. 비교

1. random 함수는 파이썬은 내장되어있지만 rust는 외부 라이브러리(크레이트?)를 사용해야한다.
2. 러스트에선 let 으로 변수 선언을 한다. mut으로 특정 함수 내에서 값이 변하는지 알린다. 변수 선언할 때 타입을 명시하지 않아도 되는 경우가 있다.(추론이 어느정도 가능하다.) 파이썬은 그런거 없다.
3. 러스트에선 (필수는 아닌듯하지만) 에러 처리가 들어가야한다. 에러가 있는 경우 .expect 등을 사용해서 처리를 한다. 파이썬 코드에선 구현하진 않았지만 일일이 찾든지 try: except: 로 에러를 처리해야한다.
4. match (condition) {value => func} 이런게 쓰인다. 파이썬에서도 (3.10 버전 이상에서) match case가 쓰인다. c의 switch 정도이다.
 